src/charm.py

```diff
@@ -278,64 +281,51 @@ def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent) -> None:
             )
             self.peer_data[self._workload_service.version] = {"openfga_model_id": openfga_model_id}
 
-        self._handle_status_update_config(event)
+        self._holistic_handler(event)
 
     def _on_openfga_store_removed(self, event: OpenFGAStoreRemovedEvent) -> None:
         if self.unit.is_leader():
             self.peer_data.pop(key=self._workload_service.version)
 
-        self._handle_status_update_config(event)
+        self._holistic_handler(event)
 
-    @wait_when(container_not_connected)
     def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        if not container_connectivity(self):
+            defer_event(event)
+
         self._workload_service.push_ca_certs(event.ca)
-        self._handle_status_update_config(event)
+        self._holistic_handler(event)
 
-    @wait_when(container_not_connected)
     def _on_certificate_removed(self, event: CertificateRemovedEvent) -> None:
-        self._workload_service.remove_ca_certs()
-        self._handle_status_update_config(event)
-
-    @wait_when(
-        container_not_connected,
-        integration_not_exists(PEER_INTEGRATION_NAME),
-    )
-    @block_when(
-        integration_not_exists(KRATOS_INFO_INTEGRATION_NAME),
-        integration_not_exists(HYDRA_ENDPOINTS_INTEGRATION_NAME),
-        integration_not_exists(OPENFGA_INTEGRATION_NAME),
-        integration_not_exists(INGRESS_INTEGRATION_NAME),
-    )
-    def _handle_status_update_config(self, event: EventBase) -> None:
-        if self.oauth_integration.is_ready() and (
-            not self.model.relations[CERTIFICATE_TRANSFER_INTEGRATION_NAME]
-        ):
-            self.unit.status = BlockedStatus(
-                "Missing certificate_transfer integration with oauth provider"
-            )
-            return
-
-        self.unit.status = MaintenanceStatus("Configuring the Admin Service container")
+        if not container_connectivity(self):
+            defer_event(event)
 
-        self._workload_service.prepare_dir(path=LOG_DIR)
-
-        if not self.openfga_integration.is_store_ready():
-            event.defer()
-            self.unit.status = WaitingStatus("Waiting for OpenFGA store")
-            return
+        self._workload_service.remove_ca_certs()
+        self._holistic_handler(event)
 
-        if not self.peer_data[self._workload_service.version]:
-            event.defer()
-            self.unit.status = WaitingStatus("Waiting for OpenFGA model")
-            return
+    def _on_collect_status(self, event: CollectStatusEvent) -> None:
+        """The central management of the charm operator's status."""
+        for condition, status in CONDITION_STATUS_REGISTRY:
+            if not condition(self):
+                event.add_status(status)
+                return
```

Comment:
Do we need to return here? I thought that we would have to add all the status in the event and let the framework decide which one to expose.

Consider this:
We have 2 failing condition status checks, one that sets the charm status to `Waiting` and one that sets it to `Blocking`. This means that we would want the charm status to be set to blocking, as that is more urgent and requires user interaction. 
With the current code, the charm would get the status of whichever condition was checked first. By not returning and adding all the status, the framework would always set the app status to `Blocked` as that has higher priority.

Reply:
Yeah, I had the same question. However, the example in the documentation and other charms that use this API seem all to do the early return. We can chat about this.